File: T1 Q5 - Asymmetric/RSA_default.cpp
#include <botan/auto_rng.h>
#include <botan/hex.h>
#include <botan/pk_keys.h>
#include <botan/pkcs8.h>
#include <botan/pubkey.h>
#include <botan/rng.h>
#include <iostream>

// Function to encrypt plaintext
std::vector<uint8_t> encrypt(const std::string &plaintext, const Botan::Public_Key &public_key, Botan::AutoSeeded_RNG &rng)
{
    std::vector<uint8_t> pt(plaintext.data(), plaintext.data() + plaintext.length());
    Botan::PK_Encryptor_EME enc(public_key, rng, "OAEP(SHA-256)");
    return enc.encrypt(pt, rng);
}

// Function to decrypt ciphertext
std::string decrypt(const std::vector<uint8_t> &ciphertext, const Botan::Private_Key &private_key, Botan::AutoSeeded_RNG &rng)
{
    Botan::PK_Decryptor_EME dec(private_key, rng, "OAEP(SHA-256)");
    Botan::secure_vector<uint8_t> pt2 = dec.decrypt(ciphertext);
    return std::string(pt2.begin(), pt2.end());
}

// Function to print hex encoded data
void print_hex(const std::string &label, const std::vector<uint8_t> &data)
{
    std::cout << label << Botan::hex_encode(data) << std::endl;
}

int main()
{
    std::string plaintext("This message is to be encrypted using RSA");
    Botan::AutoSeeded_RNG rng;

    // Load keypair
    std::string pem_file = "rsa_private.pem";
    Botan::DataSource_Stream in(pem_file);
    auto kp = Botan::PKCS8::load_key(in);

    // Separate public and private keys
    const Botan::Private_Key &private_key = dynamic_cast<const Botan::Private_Key &>(*kp);
    const Botan::Public_Key &public_key = dynamic_cast<const Botan::Public_Key &>(*kp);

    // Encrypt plaintext
    std::vector<uint8_t> ciphertext = encrypt(plaintext, public_key, rng);

    // Decrypt ciphertext
    std::string decrypted_text = decrypt(ciphertext, private_key, rng);

    // Print results
    std::cout << "_____Plaintext: " << plaintext << std::endl;
    print_hex("Encrypted text: ", ciphertext);
    std::cout << "Decrypted text: " << decrypted_text + "\n"
              << std::endl;

    return 0;
}

File: T1 Q5 - Asymmetric/RSA_default.py
import botan3 as botan

# Function to encrypt plaintext
def encrypt(plaintext, public_key):
    try:
        rng = botan.RandomNumberGenerator()
        enc = botan.PKEncrypt(public_key, "OAEP(SHA-256)")
        ciphertext = enc.encrypt(plaintext.encode('utf-8'), rng)
        return ciphertext
    except Exception as e:
        print(f"Encryption error: {e}")
        return b''

# Function to decrypt ciphertext
def decrypt(ciphertext, private_key):
    try:
        dec = botan.PKDecrypt(private_key, "OAEP(SHA-256)")
        plaintext = dec.decrypt(ciphertext).decode('utf-8')
        return plaintext
    except Exception as e:
        print(f"Decryption error: {e}")
        return ''

# Function to print hex encoded data
def print_hex(label, data):
    hex_data = data.hex()
    print(f'{label}{hex_data}')

def main():
    plaintext = "This message is to be encrypted using RSA"
    print(f"_____Plaintext: {plaintext}")

    # Generate RSA key pair
    rng = botan.RandomNumberGenerator()
    rsa_private_key = botan.PrivateKey.create('rsa', 2048, rng)
    rsa_public_key = rsa_private_key.get_public_key()

    # Encrypt plaintext
    ciphertext = encrypt(plaintext, rsa_public_key)
    print_hex("Encrypted text: ", ciphertext)

    # Decrypt ciphertext
    decrypted_text = decrypt(ciphertext, rsa_private_key)

    # Print results
    print(f"Decrypted text: {decrypted_text}\n")

if __name__ == "__main__":
    main()
